src/main/java/uk/ac/cam/cl/dtg/isaac/api/AssignmentFacade.java

```diff
@@ -939,105 +941,154 @@ public Response getAssignedGroupsByGameboards(@Context final HttpServletRequest
+                        continue;
+                    }
 
-                for (GameboardItem item : gameboard.getContents()) {
-                    if (item.getTags().contains(tag)) {
-                        this.userBadgeManager.updateBadge(currentlyLoggedInUser,
-                                UserBadgeManager.Badge.TEACHER_BOOK_PAGES_SET, assignmentWithID.getId().toString());
-                        break tagsLoop;
+                    assignmentDTO.setOwnerUserId(currentlyLoggedInUser.getId());
+                    assignmentDTO.setCreationDate(null);
+                    assignmentDTO.setId(null);
+
+                    // modifies assignment passed in to include an id.
+                    AssignmentDTO assignmentWithID = this.assignmentManager.createAssignment(assignmentDTO);
+
+                    LinkedHashMap<String, Object> eventDetails = new LinkedHashMap<>();
+                    eventDetails.put(Constants.GAMEBOARD_ID_FKEY, assignmentWithID.getGameboardId());
+                    eventDetails.put(GROUP_FK, assignmentWithID.getGroupId());
+                    eventDetails.put(ASSIGNMENT_FK, assignmentWithID.getId());
+                    eventDetails.put(ASSIGNMENT_DUEDATE_FK, assignmentWithID.getDueDate());
+                    this.getLogManager().logEvent(currentlyLoggedInUser, request, IsaacServerLogType.SET_NEW_ASSIGNMENT, eventDetails);
+
+                    this.userBadgeManager.updateBadge(currentlyLoggedInUser,
+                            UserBadgeManager.Badge.TEACHER_ASSIGNMENTS_SET, assignmentWithID.getId().toString());
+
+                    tagsLoop:
+                    for (String tag : bookTags) {
+                        for (GameboardItem item : gameboard.getContents()) {
+                            if (item.getTags().contains(tag)) {
+                                this.userBadgeManager.updateBadge(currentlyLoggedInUser,
+                                        UserBadgeManager.Badge.TEACHER_BOOK_PAGES_SET, assignmentWithID.getId().toString());
+                                break tagsLoop;
+                            }
+                        }
                     }
+                    // Assigning to this group was a success
+                    assigmentSuccessfulGroupIds.add(assignmentWithID.getGroupId());
+                } catch (DuplicateAssignmentException e) {
+                    assigmentErrors.add(new AssignmentErrorDTO(assignmentDTO.getGroupId(), e.getMessage()));
+                } catch (SegueDatabaseException e) {
+                    log.error("Database error while trying to assign work", e);
+                    assigmentErrors.add(new AssignmentErrorDTO(assignmentDTO.getGroupId(), "Unknown database error."));
                 }
             }
-
-            return Response.ok(assignmentDTOFromClient).build();
+            return Response.ok(new AssignmentSettingResponseDTO(assigmentSuccessfulGroupIds, assigmentErrors)).build();
```

Comment:
I don't think this list of group IDs is enough to correctly tell the user which assignments have been set and which have failed. Assume that all the assignments submitted are to the same group, but some are duplicates and some gamebaords don't exist; only one group ID will appear some number of times in this list, and there will be some number of errors in the other list.

Is that really enough information to inform the user what has actually happened and what has not? I've not looked at the frontend code to see the feedback it gives the user, so if it _is_ enough, then that's okay!

Reply:
I think an ordered list of statuses, one per submitted assignment, would work otherwise. (Like a list of `AssignmentStatusDTO` objects, instead of a wrapper object overall, and instead of an error object, that has `assignmentId` non-null if the assignment was set, and `errorMessage` non-null if it failed?).

But if there's enough as-is to do clear feedback, then maybe stick with this.

Reply:
Yeah having one feedback object per assignment/group pair would be better, I'll make that change tomorrow, it should take about 30 mins so it's worth it for more sensible code. 
As the frontend is currently, you can't set multiple assignments to a single group, only a single assignment to multiple groups - the error objects and list of "successful" ids tell you all you need to know for useful feedback.